### ISSUE TITLE
No force_fp32?

### ISSUE THREAD

**Issue by twmht** [open-mmlab/mmengine#550]
Hi,

Currntly I am upgrading my project to include 3.x mmcv and mmdetection.

In mmdetection 2.x, there is force_fp32 when calculating the loss (https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/dense_heads/fcos_head.py#L167).

But in mmdetection  3.x branch, there is no foce_fp32 (https://github.com/open-mmlab/mmdetection/blob/v3.0.0rc0/mmdet/models/dense_heads/fcos_head.py#L179), this would cause incorrect gradients.

the similar question can also be found in mmclassfication (https://github.com/open-mmlab/mmclassification/tree/v1.0.0rc0).

Is mmengine ready to be used?

**Comment by HAOCHENYE** [comment]
Thanks for your feedback! Since Pytorch 1.6 has supported automatic mixed precision training:[`autocast`](https://pytorch.org/docs/stable/amp.html?highlight=autocast#torch.autocast), therefore `mmengine` uses the native `autocast` to replace `force_fp16` defined in MMCV 1.0.   
The official `autocast` tutorial does not mention that we have to convert the loss to fp32, and we have verified that it will not hurt the accuracy of the downstream repo, such as MMDet, MMCLs .etc.

**Comment by twmht** [comment]
From my experience, sometimes we may get the explosion loss when using MSE loss. MES loss is usully used in distilling, which is used for two feature maps. if I don't convert the feat to fp32, then the loss would easily reached to 65536, thus we can't converage.

From my opnition I think we should keep force_fp32 utility.

**Comment by HAOCHENYE** [comment]
According to https://pytorch.org/docs/stable/amp.html?highlight=autocast#cuda-ops-that-can-autocast-to-float32, `autocast` will only convert the output of some specific torch functions to fp16, and keep the output of other functions as fp16.   The following minimum example can show this process:

```python
import torch
from torch import autocast
import torch.nn as nn


model = nn.Linear(1, 1).cuda()
with autocast(device_type='cuda'):
    x = torch.ones(1, 1).cuda()
    y = model(x)
    print(y.dtype)
    loss = nn.functional.mse_loss(y, torch.Tensor(1).cuda())
    print(loss.dtype)
```

If your custom loss function requires the input tensor must be fp32, you should convert the input data to fp32 manually before calling the loss function.

**Comment by twmht** [comment]
got it
